geoAssembler/optimiser/utility.py

```diff
@@ -0,0 +1,131 @@
+from typing import Union, Tuple
+
+
+class Integrator:
+    """
+    Object wrapping pyFAI and extra-geom geometries to probide a more
+    convenient way of integrating 2d detector images.
+
+    Creating the integrator requires a geometry, as well as the distance
+    from the sample to the detector in mm. If unknown, 200mm seems to be
+    a good 'normal' guess for AGIPD.
+
+    Once created, use the `integrate2d` method to get a 2d unrolled view
+    of the detector image. You can them sum over the image to get a 1d
+    integration result.
+    """
+
+    def __init__(self, geom: DetectorGeometryBase,
+                 sample_dist_mm: Union[int, float], unit: str = "2th_deg"):
+        self.unit = unit
+        self.sample_dist_mm = sample_dist_mm
+
+        fakedata = np.zeros(geom.expected_data_shape)
+        fakeimage, centre_geom = geom.position_modules_fast(fakedata)
+        self.size = fakeimage.shape
+
+        #  pyFAI takes in coordinates in the opposite order to those
+        #  produced by extra-geom, so swap it around here
+        self.centre = [centre_geom[1], centre_geom[0]]
+
+        self.detector = Detector(
+            pixel1=geom.pixel_size,
+            pixel2=geom.pixel_size
+        )
+
+        ai = AzimuthalIntegrator(detector=self.detector)
+        ai.setFit2D(self.sample_dist_mm, self.centre[0], self.centre[1])
+        self.ai = ai
+
+        self.radius = ((self.size[0]/2)**2 + (self.size[1]/2)**2)**(1/2)
+        self.azimuth_bins = self.radius * (self.size[0]/self.size[1])
+
+    def integrate2d(self, frame: np.ndarray, centre_offset=None):
+        """
+        Unroll the image - changes the axis from cartesian x/y to
+        polar radius/azimuthal angle.
+
+        A sum over the angle can be performed to get a result for
+        the azimuthal integration.
+
+        Returns a pyFAI `Integrate2dResult` object, check pyFAI
+        docs for more information.
+        """
+        if centre_offset is not None:
+            self.ai.setFit2D(
+                self.sample_dist_mm,
+                centre_offset[0] + self.centre[0],
+                centre_offset[1] + self.centre[1]
+            )
+
+        return self.ai.integrate2d(
+            frame,
+            self.radius,
+            self.azimuth_bins,
+            unit=self.unit,
+            dummy=np.nan,
+            method='cython'
+        )
+
+
+def avg_frame(run: DataCollection, geom: DetectorGeometryBase,
+              train_index: int, pulse_pattern=None,
+              masking=True,) -> Tuple[np.ndarray, Tuple]:
+    """
+    Returns the average value of a train with some built-in masking  of known
+    bad pixels and some corrections. Designed for use with AGIPD only.
+
+    Requires an extra-data run, extra-geom geometry, and a train index,
+    optionally can provide a slice as the pulse pattern if required.
+
+    Returns a 2d image, and the centre of the image as given by the geometry.
+    """
```

Comment:
I agree :P how do you think it should be implemented? I guess just as a property that returns ones for the asic edges, something like `geom.asic_edges`, so that users could do `assembled_image[geom.asic_edges] = np.nan`, or deal with them however they want?

As for removing bad values, makes sense, I'll take it out of the code and move it to the example with an explanation. It was in the code as if you don't try to deal with the high error values somehow then even a single pixel at a ridiculously high value breaks the optimisation, so there wasn't a real scenario where you wouldn't want to mask high values off, but it should be more explicitly stated that this is being done.

Reply:
I think I'd do it as a standalone function `agipd_asic_edges()`, returning a (512, 128) array distinguishing the edge pixels from the normal pixels. It's the same for every module, so you can apply that to a stacked array using numpy broadcasting, and then do the image assembly afterwards.